### PR TITLE
fix:fix the deadlock in EventDispatcherImpl.java

### PR DESCRIPTION
**Pull request by MlxChange**
## Summary:

[
](fix:dead lock between EventDispatcherImpl.moveStagedEventsToDispatchQueue and EventDispatcherImpl.dispatchEvent when project fabric global switch is on, bridgeLess switch is off, but the rn page set fabric is false when new RootView.)

## Changelog:
[ANDROID] [FIXED] - dead lock between EventDispatcherImpl.moveStagedEventsToDispatchQueue and EventDispatcherImpl.dispatchEvent when project fabric global switch is on, bridgeLess switch is off, but the rn page set fabric is false when new RootView.

"mqt_js" prio=5 tid=100 Blocked
  | group="main" sCount=1 ucsCount=0 flags=1 obj=0x147809e8 self=0x77155c6000
  | sysTid=22131 nice=0 cgrp=default sched=0/0 handle=0x76c9a46cb0
  | state=S schedstat=( 3880738539 230864125 1722 ) utm=364 stm=23 core=6 HZ=100
  | stack=0x76c9943000-0x76c9945000 stackSize=1039KB
  | held mutexes=
  at com.facebook.react.uimanager.events.EventDispatcherImpl.dispatchEvent(EventDispatcherImpl.java:119)
  - waiting to lock <0x01c37643> (a java.lang.Object) held by thread 1
  at com.facebook.react.fabric.interop.InteropEventEmitter.receiveEvent(InteropEventEmitter.java:50)
  at com.facebook.yrn.manager.uimodule.YRNReactEventEmitter.receiveEvent(YRNReactEventEmitter.java:97)
  at com.facebook.yrn.manager.uimodule.YRNReactEventEmitter.receiveEvent(YRNReactEventEmitter.java:74)
  at com.facebook.yrn.manager.uimodule.YRNReactEventEmitter.receiveEvent(YRNReactEventEmitter.java:66)
  at com.facebook.react.uimanager.events.Event.dispatch(Event.java:195)
  at com.facebook.react.uimanager.events.Event.dispatchModern(Event.java:240)
  at com.facebook.react.uimanager.events.EventDispatcherImpl$DispatchEventsRunnable.run(EventDispatcherImpl.java:370)
  - locked <0x0c8fcff2> (a java.lang.Object)
  at android.os.Handler.handleCallback(Handler.java:942)
  at android.os.Handler.dispatchMessage(Handler.java:99)
  at com.facebook.react.bridge.queue.MessageQueueThreadHandler.dispatchMessage(MessageQueueThreadHandler.java:27)
  at android.os.Looper.loopOnce(Looper.java:211)
  at android.os.Looper.loop(Looper.java:300)
  at com.facebook.react.bridge.queue.MessageQueueThreadImpl$4.run(MessageQueueThreadImpl.java:228)
  at java.lang.Thread.run(Thread.java:1012)

<0x0c8fcff2> -  mEventsToDispatchLock
<0x01c37643> -  mEventsStagingLock 

"main" prio=5 tid=1 Blocked
  | group="main" sCount=1 ucsCount=0 flags=1 obj=0x71de5268 self=0xb4000077f2e42c00
  | sysTid=21838 nice=-4 cgrp=default sched=0/0 handle=0x78a1d024f8
  | state=S schedstat=( 8894755690 2108474736 15229 ) utm=731 stm=157 core=6 HZ=100
  | stack=0x7fe73c1000-0x7fe73c3000 stackSize=8188KB
  | held mutexes=
  at com.facebook.react.uimanager.events.EventDispatcherImpl.moveStagedEventsToDispatchQueue(EventDispatcherImpl.java:200)
  - waiting to lock <0x0c8fcff2> (a java.lang.Object) held by thread 100
  - locked <0x01c37643> (a java.lang.Object)
  at com.facebook.react.uimanager.events.EventDispatcherImpl.-$$Nest$mmoveStagedEventsToDispatchQueue(unavailable:0)
  at com.facebook.react.uimanager.events.EventDispatcherImpl$ScheduleDispatchFrameCallback.doFrame(EventDispatcherImpl.java:291)
  at com.facebook.react.modules.core.ReactChoreographer$ReactChoreographerDispatcher.doFrame(ReactChoreographer.java:175)
  - locked <0x0913f4c0> (a java.lang.Object)
  at com.facebook.react.modules.core.ChoreographerCompat$FrameCallback$1.doFrame(ChoreographerCompat.java:85)
  at android.view.Choreographer$CallbackRecord.run(Choreographer.java:1503)
  at android.view.Choreographer$CallbackRecord.run(Choreographer.java:1513)
  at android.view.Choreographer.doCallbacks(Choreographer.java:1128)
  at android.view.Choreographer.doFrame(Choreographer.java:1037)
  at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:1481)
  at android.os.Handler.handleCallback(Handler.java:942)
  at android.os.Handler.dispatchMessage(Handler.java:99)
  at android.os.Looper.loopOnce(Looper.java:211)
  at android.os.Looper.loop(Looper.java:300)
  at android.app.ActivityThread.main(ActivityThread.java:8503)
  at java.lang.reflect.Method.invoke(Native method)
  at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:561)
  at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:954)

<0x0c8fcff2> -  mEventsToDispatchLock
<0x01c37643> -  mEventsStagingLock

### CODE REVIEW

File: packages/react-native/ReactAndroid/src/main/java/com/facebook/react/uimanager/events/EventDispatcherImpl.java [hunk 1]
```diff
@@ -118,10 +120,16 @@ public void dispatchEvent(Event event) {
...
+    try {
+      // 尝试获取事件暂存信号量，最多尝试5次，每次间隔200毫秒
+      if (tryAcquireSemaphore(mEventsStagingSemaphore, 5,200)) {
+        mEventStaging.add(event);
+        Systrace.startAsyncFlow(
           Systrace.TRACE_TAG_REACT_JAVA_BRIDGE, event.getEventName(), event.getUniqueID());
+      }
+    } finally {
+      // 释放事件暂存信号量
+      mEventsStagingSemaphore.release();
```
Comment (javache) [inline packages/react-native/ReactAndroid/src/main/java/com/facebook/react/uimanager/events/EventDispatcherImpl.java:132]: Please add comments in English

File: packages/react-native/ReactAndroid/src/main/java/com/facebook/react/uimanager/events/EventDispatcherImpl.java [hunk 2]
```diff
@@ -59,43 +61,43 @@
...
-          if (lhs == null && rhs == null) {
-            return 0;
-          }
-          if (lhs == null) {
-            return -1;
-          }
-          if (rhs == null) {
-            return 1;
-          }
+    new Comparator<Event>() {
```
Comment (javache) [inline packages/react-native/ReactAndroid/src/main/java/com/facebook/react/uimanager/events/EventDispatcherImpl.java:64]: Revert changes to whitespace please

**Comment by javache** [comment]
> when project fabric global switch is on, bridgeLess switch is off, but the rn page set fabric is false when new RootView.

This is no longer a configuration we support for open-source consumers of React Native. You must use Fabric with Bridgeless in the new architecture.

**Review by javache — CHANGES_REQUESTED** [review]

File: packages/react-native/ReactAndroid/src/main/java/com/facebook/react/uimanager/events/EventDispatcherImpl.java (hunk 2 above)
Comment (MlxChange) [reply packages/react-native/ReactAndroid/src/main/java/com/facebook/react/uimanager/events/EventDispatcherImpl.java]: it‘s fixed

File: packages/react-native/ReactAndroid/src/main/java/com/facebook/react/uimanager/events/EventDispatcherImpl.java (hunk 1 above)
Comment (MlxChange) [reply packages/react-native/ReactAndroid/src/main/java/com/facebook/react/uimanager/events/EventDispatcherImpl.java]: done

**Comment by MlxChange** [comment]
> > 当项目fabric全局开关为on时，bridgeLess开关为off，但是new RootView时rn页面设置fabric为false。
> 
> 我们不再为 React Native 的开源消费者提供此配置。您必须在新架构中将 Fabric 与 Bridgeless 结合使用。

Sorry, my description is not clear enough. It is not just this case. Some engine codes have been modified in our business, but the modified part is limited to UI synchronization. In this case, ANR problems caused by deadlocks often occur. I After observing the issues in the comment area, I found that some users have also encountered these problems, so I wonder if using semaphores instead of locks is a better solution? Finally, please forgive my not very good English, thanks

**Comment by MlxChange** [comment]
@javache Sorry to bother you, I would like to know if there is any progress on this matter?

**Comment by javache** [comment]
Sorry for the delay. I don't think we want to merge this change. Running a mix of Fabric and non-Fabric roots is not something we support anymore, and this adds risks to the legacy Paper code paths, which have significantly less internal coverage on.

cc @cipolleschi

**Pull request closed without merging**